### PR TITLE
Lev guard

### PR DESCRIPTION
**Pull request by yotamdaniel**

### CODE REVIEW

File: contracts/Coordinator.sol [hunk 1]
```diff
@@ -67,6 +68,28 @@ contract Coordinator is ICoordinator, AccessController, ReentrancyGuardUpgradeab
         _lvUSD.safeApprove(_addressPoolManager, type(uint256).max);
     }
 
+    function _coordinatorLvUSDTransferToExchanger(uint256 amount) internal {
+        uint256 currentCoordinatorLvUSDBalance = getAvailableLeverage();
```
Comment (yotamdaniel) [inline contracts/Coordinator.sol:72]: Change lvUSD to Leverage

File: contracts/Coordinator.sol [hunk 2]
```diff
@@ -67,6 +68,28 @@ contract Coordinator is ICoordinator, AccessController, ReentrancyGuardUpgradeab
...
+        require(currentCoordinatorLvUSDBalance >= amount, "insuf levAv to trnsf");
+        require(currentBalanceOnLvUSDContract >= amount, "insuf lvUSD balance to trnsf");
+
+        _paramStore.changeCoordinatorLeverageBalance(currentCoordinatorLvUSDBalance - amount);
+        _lvUSD.safeTransfer(_addressExchanger, amount);
+    }
+
+    function acceptLeverageAmount(uint256 levergaeAmountToAccept) external onlyAdmin nonReentrant {
+        uint256 currentlvUSDBalance = _lvUSD.balanceOf(address(this));
+        require(currentlvUSDBalance >= levergaeAmountToAccept, "lvUSD != levAmt");
```
Comment (yotamdaniel) [inline contracts/Coordinator.sol:83]: change to <

File: contracts/Coordinator.sol [hunk 3]
```diff
@@ -67,6 +68,28 @@ contract Coordinator is ICoordinator, AccessController, ReentrancyGuardUpgradeab
...
+        uint256 currentCoordinatorLvUSDBalance = getAvailableLeverage();
+        uint256 currentBalanceOnLvUSDContract = _lvUSD.balanceOf(address(this));
+        require(currentCoordinatorLvUSDBalance >= amount, "insuf levAv to trnsf");
+        require(currentBalanceOnLvUSDContract >= amount, "insuf lvUSD balance to trnsf");
+
+        _paramStore.changeCoordinatorLeverageBalance(currentCoordinatorLvUSDBalance - amount);
+        _lvUSD.safeTransfer(_addressExchanger, amount);
+    }
+
+    function acceptLeverageAmount(uint256 levergaeAmountToAccept) external onlyAdmin nonReentrant {
```
Comment (yotamdaniel) [inline contracts/Coordinator.sol:81]: Typo - -levergae-AmountToAccept

File: contracts/Coordinator.sol [hunk 4]
```diff
@@ -67,6 +68,28 @@ contract Coordinator is ICoordinator, AccessController, ReentrancyGuardUpgradeab
...
+        uint256 currentlvUSDBalance = _lvUSD.balanceOf(address(this));
+        require(currentlvUSDBalance >= levergaeAmountToAccept, "lvUSD != levAmt");
+        _paramStore.changeCoordinatorLeverageBalance(levergaeAmountToAccept);
+    }
+
+    function resetAndBurnLeverage() external onlyAdmin nonReentrant {
+        uint256 corrdinatorCurrentLvUSDBalance = _lvUSD.balanceOf(address(this));
+        ERC20Burnable(_addressLvUSD).burn(corrdinatorCurrentLvUSDBalance);
```
Comment (yotamdaniel) [inline contracts/Coordinator.sol:89]: Check if there is allowance to do so

File: contracts/Exchanger.sol [hunk 5]
```diff
@@ -93,6 +93,15 @@ contract Exchanger is AccessController, ReentrancyGuardUpgradeable, IExchanger,
         _index3CRV = 1;
     }
 
+    function _exhangerLvUSDTransferToCoordinator(uint256 amount) internal {
+        /// Is it possible to exploit via transferring lvUSD to exchanger which then go back to coordinator?
+        uint256 currentCoordinatorLvUSDBalance = _lvUSD.balanceOf(address(this));
```
Comment (yotamdaniel) [inline contracts/Exchanger.sol:98]: instead of _lvUSD.balance, use ParamStore LeverageBalance

Comment (yotamdaniel) [reply contracts/Exchanger.sol:98]: + change variable name to currentCoorLeverageAvaliable

File: test/EndToEndUnitTests.ts [hunk 6]
```diff
@@ -657,4 +616,101 @@ describe("Test suit for getting leverage", function () {
...
+
+        const coordinatorLvUSDBalance = await r.lvUSD.balanceOf(r.coordinator.address);
+        expect(coordinatorLvUSDBalance).to.eq(0);
+
+        const coordinatorLeverageBalance = await r.parameterStore.getCoordinatorLeverageBalance();
+        expect(coordinatorLeverageBalance).to.eq(0);
+    });
+
+    it("Should not allow any opening of position as no leverage is available", async () => {
+        await verifyPositionCreationFails("Not enough available lvUSD");
```
Comment (yotamdaniel) [inline test/EndToEndUnitTests.ts:657]: Change error message to "Not enough available leverage"

File: test/EndToEndUnitTests.ts [hunk 7]
```diff
@@ -657,4 +616,101 @@ describe("Test suit for getting leverage", function () {
...
+    });
+
+    it("Should update leverage available when taking position", async () => {
+        secondPoisitionLvUSDBorrowed = getFloatFromBigNum(await r.cdp.getLvUSDBorrowed(1));
+        const leverageAvailable = getFloatFromBigNum(await r.coordinator.getAvailableLeverage());
+        expect(newCoordinatorLvUSDBalance - secondPoisitionLvUSDBorrowed).to.closeTo(leverageAvailable, 0.0000001);
+    });
+
+    it("Should not increase available leverage if more lvUSD is minted to coordinator", async () => {
+        const beforeMintingLeverageAmount = getFloatFromBigNum(await r.coordinator.getAvailableLeverage());
```
Comment (yotamdaniel) [inline test/EndToEndUnitTests.ts:697]: change it to current lvUSD balance before minting

**Review by jasonstockman — APPROVED** [review]

**Pull request merged**